File: go/core/sentences.go
package varcore

// Splits a block of text into sentence-level spans so the matcher can try each
// sentence independently — port of sentences.ts / sentences.rs. Operates on
// runes with a running UTF-16 offset table; emitted offsets are UTF-16.

// sentence is a sentence: the trimmed text plus its UTF-16 offsets into the input.
type sentence struct {
	text        string
	startOffset int
	endOffset   int
}

var abbreviations = []string{"e.g.", "i.e.", "etc.", "cf.", "vs."}

// splitSentences splits text on ./!/?/newline terminators, skipping backtick
// code-span and double-quoted interiors, and treating decimals and a fixed
// abbreviation list as non-terminating dots.
func splitSentences(text string) []sentence {
	chars := []rune(text)
	n := len(chars)

	// Prefix table: cpToU16[i] = UTF-16 offset of char i; cpToU16[n] = total.
	cpToU16 := make([]int, n+1)
	for i := 0; i < n; i++ {
		cpToU16[i+1] = cpToU16[i] + utf16RuneLen(chars[i])
	}

	// Mark backtick code spans and double-quoted strings as no-split zones.
	skip := make([]bool, n)
	j := 0
	for j < n {
		c := chars[j]
		if c == '`' || c == '"' {
			close := findChar(chars, j+1, c)
			if close < 0 {
				break
			}
			for k := j; k <= close; k++ {
				skip[k] = true
			}
			j = close
		}
		j++
	}

	var out []sentence
	segmentStart := 0
	i := 0
	for i < n {
		if skip[i] {
			i++
			continue
		}
		ch := chars[i]
		if ch == '\n' || ch == '.' || ch == '!' || ch == '?' {
			if ch == '.' && isInsideNumberOrAbbrev(chars, i) {
				i++
				continue
			}
			end := i + 1
			out = pushSegment(out, chars, cpToU16, segmentStart, end)
			i = end
			// Skip following whitespace so the next sentence starts at content.
			for i < n && (chars[i] == ' ' || chars[i] == '\n') {
				i++
			}
			segmentStart = i
			continue
		}
		i++
	}
	out = pushSegment(out, chars, cpToU16, segmentStart, n)
	return out
}

func findChar(chars []rune, from int, target rune) int {
	for k := from; k < len(chars); k++ {
		if chars[k] == target {
			return k
		}
	}
	return -1
}

func pushSegment(out []sentence, chars []rune, cpToU16 []int, start, end int) []sentence {
	if end <= start {
		return out
	}
	raw := string(chars[start:end])
	slice := javaStrip(raw)
	if slice == "" {
		return out
	}
	leading := utf16Len(raw) - utf16Len(javaStripLeading(raw))
	trimmedStart := cpToU16[start] + leading
	trimmedEnd := trimmedStart + utf16Len(slice)
	return append(out, sentence{
		text:        slice,
		startOffset: trimmedStart,
		endOffset:   trimmedEnd,
	})
}

func isInsideNumberOrAbbrev(chars []rune, dotPos int) bool {
	var prev, next rune
	if dotPos > 0 {
		prev = chars[dotPos-1]
	}
	if dotPos+1 < len(chars) {
		next = chars[dotPos+1]
	}
	if isASCIIDigit(prev) && isASCIIDigit(next) {
		return true
	}
	// Known abbreviations ending at dotPos+1.
	for _, abbrev := range abbreviations {
		length := len([]rune(abbrev))
		from := dotPos + 1 - length
		if from < 0 {
			from = 0
		}
		candidate := string(chars[from : dotPos+1])
		if candidate == abbrev {
			return true
		}
	}
	// Lowercase letter following → likely intra-word.
	return next >= 'a' && next <= 'z'
}

func isASCIIDigit(r rune) bool { return r >= '0' && r <= '9' }

File: rust/core/src/sentences.rs
//! Splits a block of text into sentence-level spans so the matcher can try each
//! sentence independently — port of `sentences.ts` / `Sentences.java`. Operates
//! on `char`s with a running UTF-16 offset table (the Python-port approach): the
//! split decisions use BMP terminators, and emitted offsets are UTF-16.

use crate::offsets::{java_strip, java_strip_leading, utf16_len};

/// A sentence: the trimmed text plus its UTF-16 offsets into the input.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Sentence {
    pub text: String,
    pub start_offset: usize,
    pub end_offset: usize,
}

impl Sentence {
    pub fn new(text: impl Into<String>, start_offset: usize, end_offset: usize) -> Sentence {
        Sentence {
            text: text.into(),
            start_offset,
            end_offset,
        }
    }
}

const ABBREVIATIONS: [&str; 5] = ["e.g.", "i.e.", "etc.", "cf.", "vs."];

/// Splits `text` on `.`/`!`/`?`/newline terminators, skipping backtick code-span
/// and double-quoted interiors, and treating decimals and a fixed abbreviation
/// list as non-terminating dots.
pub fn split_sentences(text: &str) -> Vec<Sentence> {
    let chars: Vec<char> = text.chars().collect();
    let n = chars.len();

    // Prefix table: cp_to_u16[i] = UTF-16 offset of char i; cp_to_u16[n] = total.
    let mut cp_to_u16 = vec![0usize; n + 1];
    for i in 0..n {
        cp_to_u16[i + 1] = cp_to_u16[i] + chars[i].len_utf16();
    }

    // Mark backtick code spans and double-quoted strings as no-split zones.
    let mut skip = vec![false; n];
    let mut j = 0;
    while j < n {
        let c = chars[j];
        if c == '`' || c == '"' {
            match find_char(&chars, j + 1, c) {
                Some(close) => {
                    for entry in skip.iter_mut().take(close + 1).skip(j) {
                        *entry = true;
                    }
                    j = close;
                }
                None => break,
            }
        }
        j += 1;
    }

    let mut out = Vec::new();
    let mut segment_start = 0usize;
    let mut i = 0;
    while i < n {
        if skip[i] {
            i += 1;
            continue;
        }
        let ch = chars[i];
        if ch == '\n' || ch == '.' || ch == '!' || ch == '?' {
            if ch == '.' && is_inside_number_or_abbrev(&chars, i) {
                i += 1;
                continue;
            }
            let end = i + 1;
            push_segment(&mut out, &chars, &cp_to_u16, segment_start, end);
            i = end;
            // Skip following whitespace so the next sentence starts at content.
            while i < n && (chars[i] == ' ' || chars[i] == '\n') {
                i += 1;
            }
            segment_start = i;
            continue;
        }
        i += 1;
    }
    push_segment(&mut out, &chars, &cp_to_u16, segment_start, n);
    out
}

fn find_char(chars: &[char], from: usize, target: char) -> Option<usize> {
    (from..chars.len()).find(|&k| chars[k] == target)
}

fn push_segment(
    out: &mut Vec<Sentence>,
    chars: &[char],
    cp_to_u16: &[usize],
    start: usize,
    end: usize,
) {
    if end <= start {
        return;
    }
    let raw: String = chars[start..end].iter().collect();
    let slice = java_strip(&raw);
    if slice.is_empty() {
        return;
    }
    let leading = utf16_len(&raw) - utf16_len(java_strip_leading(&raw));
    let trimmed_start = cp_to_u16[start] + leading;
    let trimmed_end = trimmed_start + utf16_len(slice);
    out.push(Sentence {
        text: slice.to_string(),
        start_offset: trimmed_start,
        end_offset: trimmed_end,
    });
}

fn is_inside_number_or_abbrev(chars: &[char], dot_pos: usize) -> bool {
    let prev = if dot_pos > 0 {
        chars[dot_pos - 1]
    } else {
        '\0'
    };
    let next = if dot_pos + 1 < chars.len() {
        chars[dot_pos + 1]
    } else {
        '\0'
    };
    if prev.is_ascii_digit() && next.is_ascii_digit() {
        return true;
    }
    // Known abbreviations ending at dot_pos+1.
    for abbrev in ABBREVIATIONS {
        let len = abbrev.chars().count();
        let from = (dot_pos + 1).saturating_sub(len);
        let candidate: String = chars[from..dot_pos + 1].iter().collect();
        if candidate == abbrev {
            return true;
        }
    }
    // Lowercase letter following → likely intra-word.
    next.is_ascii_lowercase()
}
